src/python/pants/backend/python/goals/package_pex_binary.py

```diff
@@ -56,7 +56,10 @@
 class PexBinaryFieldSet(PackageFieldSet, RunFieldSet):
     run_in_sandbox_behavior = RunInSandboxBehavior.RUN_REQUEST_HERMETIC
 
-    required_fields = (PexEntryPointField,)
+    required_fields = (
+        PexEntryPointField,
+        PexScriptField,
```

Comment:
This may be just words getting in the way, but those fields are mutually exclusive; so to have them both "required" means _something_ is off.

Reply:
Ah. `required_fields` means "the fields that are _defined_ on the target that will cause the `FieldSet` magic to match this field". It has nothing to do with the values being required, it has to do with the fields themselves being required.

Reply:
Still a bit weird since it implies you'd actually work just fine with a Target that was defined with just one of the two fields, but not both. Something definitely is a bit broken there in the pre-existing magic / semantics.

Reply:
Uhm, I still don't see why you had to add the `PexScriptField` to the list of `required_fields` here.  The "required" part is indeed for identifying eligible targets for a particular field set, but apart from that, all the magic happens for all the fields of the target onto the field set any way (i.e. there's no need to list all fields referenced by the field set in the list of required fields).

Example demonstrating my point: 

https://github.com/pantsbuild/pants/blob/9abc93794ceee0c0550666b8dff0e0b393f4c72e/src/python/pants/backend/docker/goals/package_image.py#L70-L80

Reply:
It created ambiguity with the new `PythonRequirementFieldSet`

Reply:
Oh, wait ok now I think I see.. is it to not have `python_requirement` targets be eligible for the `PexBinaryFieldSet`?

Reply:
That makes sense and deserves a comment or just using that field set. The infra issue still remains. If a rule needs one field or another or both, it should be able to say that.

Reply:
The problem was that because the _only_ field on `PexBinaryFieldSet` was `PexEntryPointField`, _any_ target that had `PexEntryPointField` created ambiguity when attempting to find the matching `FieldSet` type. Generally speaking, `required_fields` should be as specific as possible, and should describe every member variable that comes from a `Field`.

Reply:
Why not also use PexBinaryFieldSet?

Reply:
Because it's easier to use Pants' built-in matching code to decide which `RunRequest` rule to execute than to re-write our own matching logic.

Reply:
You're basically paying a high price in review (as are we) for not commenting a sketchy move forced upon you by infra as it stands.

Reply:
I wasn't even aware there could be ambiguity issues wrgt field sets... 🤔

Reply:
Yeah. I think we should probably look at renaming `required_fields` in a version or two. There's clearly some conceptual confusion between it and `Field.required`, even though there is no link between them.

Reply:
> I wasn't even aware there could be ambiguity issues wrgt field sets... 🤔

Yeah, because most field sets depend on 3+ fields, which makes them reasonably unique.

Reply:
This ambiguity issue isn't mentioned here: https://www.pantsbuild.org/docs/rules-api-and-target-api#fieldsets which feels like an omission.

Reply:
I think the real solution here is that `python_requirement` shouldn't be using a `PexEntryPointField`, but rather it's own subclass. `PythonRequirementEntryPointField`.

This happened a while ago to `Dependencies`.